packages/runtime-common/index-writer.ts

```diff
@@ -512,31 +512,7 @@ export class Batch {
     }
   }
 
-  async invalidate(urls: URL[]): Promise<string[]> {
-    await this.ready;
-    let start = Date.now();
-    this.#perfLog.debug(
-      `${jobIdentity} starting invalidation of ${urls.map((u) => u.href).join()}`,
-    );
-    let visited = new Set<string>();
-    let invalidations: string[] = [];
-    for (let url of urls) {
-      let alias = trimExecutableExtension(url).href;
-      let workingInvalidations = [
-        ...new Set([
-          ...(!this.nodeResolvedInvalidations.includes(alias)
-            ? [url.href]
-            : []),
-          ...(alias ? await this.calculateInvalidations(alias, visited) : []),
-        ]),
-      ];
-      invalidations = [...new Set([...invalidations, ...workingInvalidations])];
-    }
-
-    if (invalidations.length === 0) {
-      return [];
-    }
-
+  async tombstoneEntries(invalidations: string[]) {
```

Comment:
can you hoist this into the constructor. we lose cohesion when we start doing side effects like this.  it means you need to add a little extra logic to set the `this.ready`, but it will make for less surprising code.

Reply:
I see your point about cohesion. 

How do I know what to tombstone before calling `.invalidate`?  Previously, this side effect of tombstoning was called inside of `invalidate` -- I guess for good reason. 

By putting tombstoning into the constructor, the question then becomes "am I just tombstoning everything at every batch run?" 

If we want this side-effect to be manually run by the current run operation. I think it makes sense to manually call `tombstoneEntries` from outside -- typically after every call to `invalidate`

Reply:
But only for from-scratch. The normal invalidate for incremental should still tombstone from the inside as part of calling invalidate()

Reply:
(I’m away from my keyboard right now and don’t remember if we call Batch.invalidate() directly for from scratch indexing). I’m assuming we don’t which is why it’s a side effect of incrementing the realm version. It should VERY much be part of the Batch.invalidate() method. I’m not suggesting removing it there. Tombstoning is part of the invalidation process, and it doing it in that method does not lead to lack of cohesion, and it’s not surprising that it happens there.

Reply:
> Sorry for all the replies to this, I think I finally understand why you did this: it’s because I said we should invalidate everything in from scratch. In hindsight that wrong because we added the mtime checks to see what has actually changed. If this tombstoning as part of the increasing the realm version is a reaction to that, then I think we can just remove it. The fact that we added the files that were in the index but not on disk to the invalidation list should be all that is required I believe, as that is technically something that has “changed”.

I believe that is all I have done. I think what is confusing is that I exposed a new tombstoneEntries method which actually should be private to the batch (now its public)